spring-integration-core/src/main/java/org/springframework/integration/transformer/FromProtobufTransformer.java

```diff
@@ -0,0 +1,173 @@
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.integration.transformer;
+
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.lang.reflect.Method;
+import java.util.Map;
+
+import com.google.protobuf.Message.Builder;
+
+import org.springframework.beans.factory.BeanClassLoaderAware;
+import org.springframework.expression.EvaluationContext;
+import org.springframework.expression.Expression;
+import org.springframework.integration.context.IntegrationContextUtils;
+import org.springframework.integration.expression.FunctionExpression;
+import org.springframework.integration.transformer.support.ProtoHeaders;
+import org.springframework.messaging.Message;
+import org.springframework.util.Assert;
+import org.springframework.util.ClassUtils;
+import org.springframework.util.ConcurrentReferenceHashMap;
+
+/**
+ * An Protocol Buffer transformer to instantiate {@link com.google.protobuf.Message} objects from {@code byte[]}.
+ * 
+ * @author Christian Tzolov
+ * @since 6.1
+ */
+public class FromProtobufTransformer extends AbstractTransformer implements BeanClassLoaderAware {
+
+	private static final Map<Class<?>, Method> methodCache = new ConcurrentReferenceHashMap<>();
+
+	private final Class<? extends com.google.protobuf.Message> defaultType;
+
+	private ClassLoader beanClassLoader;
+
+	private Expression typeIdExpression = new FunctionExpression<Message<?>>(
+			(message) -> message.getHeaders().get(ProtoHeaders.TYPE));
+
+	private EvaluationContext evaluationContext;
+
+	/**
+	 * Construct an instance with the supplied default type to create.
+	 * @param defaultType the type.
+	 */
+	public FromProtobufTransformer(Class<? extends com.google.protobuf.Message> defaultType) {
+		Assert.notNull(defaultType, "'defaultType' must not be null");
+		this.defaultType = defaultType;
+	}
+
+	@Override
+	public void setBeanClassLoader(ClassLoader classLoader) {
+		this.beanClassLoader = classLoader;
+	}
+
+	/**
+	 * Set the expression to evaluate against the message to determine the type. Default {@code headers['proto_type']}.
+	 * @param expression the expression.
+	 * @return the transformer
+	 */
+	public FromProtobufTransformer typeExpression(Expression expression) {
+		assertExpressionNotNull(expression);
+		this.typeIdExpression = expression;
+		return this;
+	}
+
+	/**
+	 * Set the expression to evaluate against the message to determine the type id. Default
+	 * {@code headers['proto_type']}.
+	 * @param expression the expression.
+	 * @return the transformer
+	 */
+	public FromProtobufTransformer typeExpression(String expression) {
```

Comment:
I think we don't need extra methods for builder style.
Can't it be just a single `public FromProtobufTransformer setTypeExpression(Expression expression) {`?
This way we will satisfy XML configuration support and will have a way to chain methods in a builder style.